### PR TITLE
WIP: Avoid stripping attributes in dplyr_reconstruct()

### PR DESCRIPTION
**Pull request by krlmlr**
Reverts #5277.

This avoids materialization of the lazy relational object.

We don't need this if https://github.com/duckdblabs/duckplyr/commit/9ceb8157d1cc2e17e1f27b6b09afd5fdc58670b7 is deemed okay.

### CODE REVIEW

**Comment by krlmlr** [comment]
@DavisVaughan: The following reprex illustrates the problem.

With stock `dplyr_reconstruct()` which is currently stripping attributes, data frames will be materialized prematurely.

``` r
con <- DBI::dbConnect(duckdb::duckdb())
df <- data.frame(a = 1)

rel1 <- duckdb:::rel_from_df(con, df)
rel2 <- duckdb:::rel_project(
  rel1,
  list(duckdb:::expr_reference("a"))
)

df_out <- duckdb:::rel_to_altrep(rel2)

options(duckdb.materialize_message = TRUE)
invisible(dplyr::dplyr_reconstruct(df_out, df))
#> materializing:
#> ---------------------
#> --- Relation Tree ---
#> ---------------------
#> Projection [a as a]
#>   r_dataframe_scan(0x12f8a9ea0)
#> 
#> ---------------------
#> -- Result Columns  --
#> ---------------------
#> - a (DOUBLE)

df_out
#>   a
#> 1 1
```

<sup>Created on 2023-05-17 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

We want the "materializing:" message to happen only when `df_out` is printed, and this is what happens if we don't call `dplyr_reconstruct()` :

``` r
con <- DBI::dbConnect(duckdb::duckdb())
df <- data.frame(a = 1)

rel1 <- duckdb:::rel_from_df(con, df)
rel2 <- duckdb:::rel_project(
  rel1,
  list(duckdb:::expr_reference("a"))
)

df_out <- duckdb:::rel_to_altrep(rel2)

options(duckdb.materialize_message = TRUE)
df_out
#> materializing:
#> ---------------------
#> --- Relation Tree ---
#> ---------------------
#> Projection [a as a]
#>   r_dataframe_scan(0x111aaf660)
#> 
#> ---------------------
#> -- Result Columns  --
#> ---------------------
#> - a (DOUBLE)
#> 
#>   a
#> 1 1
```

<sup>Created on 2023-05-17 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

This message indicates when the actual computation is happening. Earlier I took some time to understand what's triggering this materialization, it's buried very deeply in vctrs. (Something related to the decision if we have named rows or not.) I figure that not stripping attributes could be the easier fix, but perhaps you have other suggestions.

To sum up: my goal is the following behavior:

``` r
con <- DBI::dbConnect(duckdb::duckdb())
df <- data.frame(a = 1)

rel1 <- duckdb:::rel_from_df(con, df)
rel2 <- duckdb:::rel_project(
  rel1,
  list(duckdb:::expr_reference("a"))
)

df_out <- duckdb:::rel_to_altrep(rel2)

options(duckdb.materialize_message = TRUE)
invisible(dplyr::dplyr_reconstruct(df_out, df))

df_out
#> materializing:
#> ---------------------
#> --- Relation Tree ---
#> ---------------------
#> Projection [a as a]
#>   r_dataframe_scan(0x12f8a9ea0)
#> 
#> ---------------------
#> -- Result Columns  --
#> ---------------------
#> - a (DOUBLE)
#>
#>   a
#> 1 1
```

<sup>Created on 2023-05-17 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by krlmlr** [comment]
I can create a small package with a self-contained demo if needed, I don't know how to create an ALTREP data type in a reprex.
